_posts/2020-08-25-googlectf-2020-root-power.markdown

```diff
@@ -0,0 +1,175 @@
+
+## The filesystem
+
+The filesystem can be mounted with:
+
+```bash
+mkdir -p mnt && sudo mount -o loop,offset=1048576 disk.img mnt
+```
+
+My first instinct was to check `/etc/shadow` and the boot files, the former was a rabbit hole.
+I also checked `initramfs` to see if there are any files I might've missed, it contains `ssdt.aml`, we'll come back to this later.
+
+## Authentication mechanism
+A feature that authenticated users in Linux was PAM.
+So I checked out the modules in `/etc/pam.d/system-auth`, and `pam_chck.so` was mentioned.
+
+`pam_chck.so` looks for the __root__ user, then returns 0 (success) if `check_device() == 1`.
+
+```c
+ulong pam_sm_authenticate(undefined8 param_1)
+{
+  int r;
+  ulong ret;
+  char *user;
+  uint _r;
+  
+  _r = pam_get_user(param_1,&user,"Username: ",&user);
+  if (_r == 0) {
+    r = strcmp(user,"root");
+    if (r == 0) {
+      fwrite("Password: ",1,10,stderr);
+      r = check_device();
+      if (r == 1) {
+        fprintf(stderr,"\n\nWelcome %s\n",user);
+        ret = 0;
+      }
+      else {
+        fwrite("Wrong username or password",1,0x1a,stderr);
+        ret = 6;
+      }
+    }
+    else {
+      ret = 6;
+    }
+  }
+  else {
+    ret = (ulong)_r;
+  }
+  return ret;
+}
+```
+
+`check_device()` is a simple function that reads 2 bytes from `/dev/chck` and returns the result as an integer.
+
+```c
+int check_device(void)
+{
+  int ret;
+  char buf [2];
+  FILE *fd;
+  
+  fd = fopen("/dev/chck","r");
+  fgets(buf,2,fd);
+  fclose(fd);
+  ret = atoi(buf);
+  return ret;
+}
+```
+
+What exactly provides this device?
+
+
+## Chck kernel module
+
+Devices in `/dev/` are usually added through kernel modules, so I looked for `.ko` files in the image.
+Immediately, `chck.ko` jumped out so it was given the decompiler treatment, indeed it handles `/dev/chck`.
+
+```c
+undefined8 chck_read(undefined8 param_1,undefined8 param_2,undefined8 param_3,undefined8 param_4)
+{
+  int chck_r;
+  undefined8 r;
+  undefined8 extraout_RDX;
+  long in_GS_OFFSET;
+  undefined8 acpi_data;
+  char local_22 [2];
+  long canary;
+  
+  __fentry__();
+  canary = *(long *)(in_GS_OFFSET + 0x28);
+  chck_r = acpi_evaluate_integer(chck_handle,"CHCK",0,&acpi_data);
+  if (chck_r == 0) {
+    snprintf(local_22,2,"%llu",acpi_data);
+    r = simple_read_from_buffer(param_2,extraout_RDX,param_4,local_22,2);
+  }
+  else {
+    printk("\x014Chck: cannot read from method CHCK");
+    r = 0xffffffffffffffff;
+  }
+  if (canary == *(long *)(in_GS_OFFSET + 0x28)) {
+    return r;
+  }
+                    // WARNING: Subroutine does not return
+  __stack_chk_fail();
+}
+```
+
+This function relays reads on `/dev/chck` to ACPI's _CHCK_ device.
+My guess was that this was handled early in the boot process --- `initramfs`.
+
+## AML file
+
+`initramfs` contains `ssdt.aml` which can be decompiled with `iasl`, producing ssdt.dsl.
+The following helped me make sense of what is happening:
+
+1. https://uefi.org/specifications
```

Comment:
Make these links?

Reply:
Didn't notice jekyll didn't do that automatically